tests/unit/test_wait_for_idle.py

```diff
@@ -0,0 +1,540 @@
+    app["units"]["hexanator/1"] = copy.deepcopy(app["units"]["hexanator/0"])
+
+    kwargs["apps"] = ["hexanator"]
+    kwargs["wait_for_exact_units"] = 2
+
+    idle = await model_fake(full_status_response)._check_idle(**kwargs)
+    assert idle
+
+
+@pytest.mark.parametrize("wrong_count", [1, 3])
+async def test_wrong_exact_count(
+    full_status_response: dict[str, Any], kwargs: dict[str, Any], wrong_count: int
+):
+    app = full_status_response["response"]["applications"]["hexanator"]
+    app["units"]["hexanator/1"] = copy.deepcopy(app["units"]["hexanator/0"])
+
+    kwargs["apps"] = ["hexanator"]
+    kwargs["wait_for_exact_units"] = wrong_count
+
+    idle = await model_fake(full_status_response)._check_idle(**kwargs)
+    assert not idle
+
+
+async def test_exact_count_precedence(
+    full_status_response: dict[str, Any], kwargs: dict[str, Any]
+):
+    app = full_status_response["response"]["applications"]["hexanator"]
+    app["units"]["hexanator/1"] = copy.deepcopy(app["units"]["hexanator/0"])
+
+    kwargs["apps"] = ["hexanator"]
+    kwargs["wait_for_exact_units"] = 2
+    kwargs["wait_for_at_least_units"] = 9
+    kwargs["_wait_for_units"] = 9
+
+    idle = await model_fake(full_status_response)._check_idle(**kwargs)
+    assert idle
+
+
+async def test_ready_unit_count(
+    full_status_response: dict[str, Any], kwargs: dict[str, Any]
+):
+    app = full_status_response["response"]["applications"]["hexanator"]
+    app["units"]["hexanator/1"] = copy.deepcopy(app["units"]["hexanator/0"])
+
+    kwargs["apps"] = ["hexanator"]
+    kwargs["wait_for_at_least_units"] = 2
+    kwargs["_wait_for_units"] = 2
+    kwargs["status"] = "active"
+
+    idle = await model_fake(full_status_response)._check_idle(**kwargs)
+    assert idle
+
+
+async def test_ready_unit_requires_idle_agent(
+    full_status_response: dict[str, Any], kwargs: dict[str, Any]
+):
+    app = full_status_response["response"]["applications"]["hexanator"]
+    app["units"]["hexanator/1"] = copy.deepcopy(app["units"]["hexanator/0"])
+    app["units"]["hexanator/1"]["agent-status"]["status"] = "some-other"
+
+    kwargs["apps"] = ["hexanator"]
+    kwargs["wait_for_at_least_units"] = 2
+    kwargs["_wait_for_units"] = 2
+    kwargs["status"] = "active"
+
+    idle = await model_fake(full_status_response)._check_idle(**kwargs)
+    assert not idle
+
+
+async def test_ready_unit_requires_workload_status(
+    full_status_response: dict[str, Any], kwargs: dict
+):
+    app = full_status_response["response"]["applications"]["hexanator"]
+    app["units"]["hexanator/1"] = copy.deepcopy(app["units"]["hexanator/0"])
+    app["units"]["hexanator/1"]["workload-status"]["status"] = "some-other"
+
+    kwargs["apps"] = ["hexanator"]
+    kwargs["wait_for_at_least_units"] = 2
+    kwargs["_wait_for_units"] = 2
+    kwargs["status"] = "active"
+
+    idle = await model_fake(full_status_response)._check_idle(**kwargs)
+    assert not idle
+
+
+# FIXME eventual timeout
```

Comment:
Not clear to me what this is referring to

Reply:
moved and expanded the list of missing tests